Loop for newest version in appstore response

LukasReschke opened the pull request:
The current implementation when fetching apps from the appstore is to assume that the first element is the newest version, this is now always applicable (e.g. daily builds etc.) and leads to the fact that for some apps (e.g. nextant) the newest version is not delivered. This can be easily tested by comparing the version of the downloaded Nextant version, without this patch it only delivers 0.6.6 instead of 0.10.0. (Category: Files)

This change will loop over all releases delivered by the appstore and chooses the newest compatible one. While not the cleanest solution, it does its job.

Most of the code are actually unit tests. Whereas I have copied the whole original response from the appstore and also have performed the transformation. So that's why the diff looks so huge.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke committed:
Loop for newest version in appstore response

The current implementation when fetching apps from the appstore is to assume that the first element is the newest version, this is now always applicable and leads to the fact that for some apps (e.g. nextant) the newest version is not delivered. This can be easily tested by comparing the version of the downloaded Nextant version.

This change will loop over all releases delivered by the appstore and chooses the newest compatible one. While not the cleanest solution, it does its job.

Most of the code are actually unit tests. Whereas I have copied the whole original response from the appstore and also have performed the transformation. So that's why the diff looks so huge.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

nickvergessen commented:
Works, afterwards i am suggested an update to 0.10.0 on the category page, normal enabled page however does not.

LukasReschke committed:
Also show updates on the "enabled" page

The enabled page doesn't pass through "getAppsForCategory" thus it also needs to have that special logic applied.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke commented:
> normal enabled page however does not.

Good point. Adjusted the logic to also cover that. This controller file is really a little bit … special …

nickvergessen commented:
Works now 👍

blizzz commented:
Code looks OK, works 👍

BernhardPosselt commented:
Hm, this does not filter pre-releases, is this what you want? Users will be prompted to upgrade to 2.0.0-beta.1 for instance

LukasReschke commented:
> Hm, this does not filter pre-releases, is this what you want? Users will be prompted to upgrade to 2.0.0-beta.1 for instance

https://github.com/nextcloud/server/pull/2307/files#diff-83840d74b03720feef2006e83453333eR83 should remove all nightlies. Am I missing something?

BernhardPosselt commented on lib/private/App/AppStore/Fetcher/AppFetcher.php:
```diff
@@ -53,4 +58,66 @@ public function __construct(IAppData $appData,
+				// Exclude all nightly releases
+				if($release['isNightly'] === false) {
+					// Exclude all versions not compatible with the current version
+					$versionParser = new VersionParser();
```
This should not be needed. All apps fetched from the API are working on the version provided in your URL

BernhardPosselt commented:
@LukasReschke right, however there are 3 "channels":
* stable (e.g. 1.0.0)
* pre-release (e.g. 1.0.0-beta)
* nightly (e.g. 1.0.0-beta + nightly flag)

LukasReschke commented:
So for what does one have to filter? The docs doesn't tell anything about pre-releases etc…

(also it seems rather unnecessary for me at the moment since Nextcloud only supports the stable channel at the moment, but fair enough…)

BernhardPosselt commented:
Pre-releases in semver are denoted by having a - after the version string. Since we don't allow build metadata in our version you can simply check with 

```php

if (strpos($release['version'], '-') === false) {
    // not a pre-release
}

```

BernhardPosselt commented:
Versioning is documented at http://nextcloudappstore.readthedocs.io/en/latest/restapi.html#get-all-apps-and-releases 👍

version
    A semantic version without build metadata (e.g. 1.3.0, 1.2.1-alpha.1)

If you think this was not enough docs to handle this case, can you create a pull request to the docs that specifically adds a short description about pre-releases :D?

nickvergessen commented:
@BernhardPosselt Ever heard of -PL releases? just saying

LukasReschke commented:
https://github.com/nextcloud/server/pull/2327 should make you all happier… I hope…

BernhardPosselt commented:
@nickvergessen nope, havent heard of that one
